src/vkscript/executor_test.cc

```diff
@@ -249,49 +279,212 @@ class VkScriptExecutorTest : public testing::Test {
+  EXPECT_EQ(FormatType::kR32G32B32A32_SFLOAT,
+            color_frame_format->GetFormatType());
+  EXPECT_EQ(FormatType::kD24_UNORM_S8_UINT,
+            depth_frame_format->GetFormatType());
+}
+
+TEST_F(VkScriptExecutorTest, ExecutesRequiredFenceTimeout) {
+  std::string input = R"(
+[require]
+fence_timeout 12345)";
+
+  Parser parser;
+  ASSERT_TRUE(parser.Parse(input).IsSuccess());
+
+  auto engine = MakeAndInitializeEngine(parser);
+
+  Executor ex;
+  Result r = ex.Execute(engine.get(), parser.GetScript());
+  ASSERT_TRUE(r.IsSuccess());
+
+  const auto& features = ToStub(engine.get())->GetFeatures();
+  ASSERT_EQ(1U, features.size());
+  EXPECT_EQ(Feature::kFenceTimeout, features[0]);
+
+  const auto& extensions = ToStub(engine.get())->GetExtensions();
+  ASSERT_EQ(0U, extensions.size());
+
+  EXPECT_EQ(12345U, ToStub(engine.get())->GetFenceTimeoutMs());
+
+  const auto* color_frame_format = ToStub(engine.get())->GetColorFrameFormat();
+  const auto* depth_frame_format = ToStub(engine.get())->GetDepthFrameFormat();
+  EXPECT_EQ(nullptr, color_frame_format);
+  EXPECT_EQ(nullptr, depth_frame_format);
+}
+
+TEST_F(VkScriptExecutorTest, ExecutesRequiredAll) {
+  std::string input = R"(
+[require]
+robustBufferAccess
+logicOp
+VK_KHR_storage_buffer_storage_class
+VK_KHR_variable_pointers
+framebuffer R32G32B32A32_SFLOAT
+depthstencil D24_UNORM_S8_UINT
+fence_timeout 12345)";
+
+  Parser parser;
+  ASSERT_TRUE(parser.Parse(input).IsSuccess());
+
+  auto engine = MakeAndInitializeEngine(parser);
 
   Executor ex;
   Result r = ex.Execute(engine.get(), parser.GetScript());
   ASSERT_TRUE(r.IsSuccess());
 
-  auto requirements = ToStub(engine.get())->GetRequirements();
-  ASSERT_EQ(1U, requirements.size());
-  EXPECT_EQ(Feature::kFramebuffer, requirements[0].feature);
-  EXPECT_EQ(FormatType::kR32G32B32A32_SINT,
-            requirements[0].format->GetFormatType());
+  const auto& features = ToStub(engine.get())->GetFeatures();
+  ASSERT_EQ(5U, features.size());
+  EXPECT_EQ(Feature::kRobustBufferAccess, features[0]);
+  EXPECT_EQ(Feature::kLogicOp, features[1]);
+  EXPECT_EQ(Feature::kFramebuffer, features[2]);
+  EXPECT_EQ(Feature::kDepthStencil, features[3]);
+  EXPECT_EQ(Feature::kFenceTimeout, features[4]);
+
+  const auto& extensions = ToStub(engine.get())->GetExtensions();
+  ASSERT_EQ(2U, extensions.size());
+  EXPECT_EQ("VK_KHR_storage_buffer_storage_class", extensions[0]);
+  EXPECT_EQ("VK_KHR_variable_pointers", extensions[1]);
+
+  EXPECT_EQ(12345U, ToStub(engine.get())->GetFenceTimeoutMs());
+
+  const auto* color_frame_format = ToStub(engine.get())->GetColorFrameFormat();
+  const auto* depth_frame_format = ToStub(engine.get())->GetDepthFrameFormat();
+  EXPECT_EQ(FormatType::kR32G32B32A32_SFLOAT,
+            color_frame_format->GetFormatType());
+  EXPECT_EQ(FormatType::kD24_UNORM_S8_UINT,
+            depth_frame_format->GetFormatType());
+}
+
+TEST_F(VkScriptExecutorTest, ExecuteDuplicatedRequirementsFailed) {
```

Comment:
I'd remove this test as it's just testing the Stub and not real engine code.

Reply:
done